relayer/store/store.go

```diff
@@ -0,0 +1,190 @@
+	InitialVerificationTxConfirmed Status = iota // 2
+	ReadyToComplete                Status = iota // 3
+	CompleteVerificationTxSent     Status = iota // 4
+)
+
+type BeefyItem struct {
+	gorm.Model
+	ValidatorAddresses         []byte
+	SignedCommitment           []byte
+	Status                     Status
+	InitialVerificationTxHash  common.Hash
+	CompleteOnBlock            uint64
+	RandomSeed                 common.Hash
+	CompleteVerificationTxHash common.Hash
+}
+
+func NewBeefyItem(validatorAddresses, signedCommitment []byte, status Status,
+	initialVerificationTxHash common.Hash, completeOnBlock uint64, randomSeed,
+	completeVerificationTxHash common.Hash) BeefyItem {
+	return BeefyItem{
+		ValidatorAddresses:         validatorAddresses,
+		SignedCommitment:           signedCommitment,
+		Status:                     status,
+		InitialVerificationTxHash:  initialVerificationTxHash,
+		CompleteOnBlock:            completeOnBlock,
+		RandomSeed:                 randomSeed,
+		CompleteVerificationTxHash: completeVerificationTxHash,
+	}
+}
+
+func (b *BeefyItem) ToBeefy() (Beefy, error) {
+	var validatorAddresses []common.Address
+	if err := json.Unmarshal(b.ValidatorAddresses, &validatorAddresses); err != nil {
+		return Beefy{}, err
+	}
+
+	var signedCommitment SignedCommitment
+	if err := json.Unmarshal(b.SignedCommitment, &signedCommitment); err != nil {
+		return Beefy{}, err
+	}
+
+	return Beefy{
+		ValidatorAddresses: validatorAddresses,
+		SignedCommitment:   signedCommitment,
+	}, nil
+}
+
+func (BeefyItem) TableName() string {
+	return "beefy_item"
+}
+
+type CmdType int
+
+const (
+	Create CmdType = iota // 0
+	Update CmdType = iota // 1
+)
+
+type DatabaseCmd struct {
+	Item         *BeefyItem
+	Type         CmdType
+	Instructions map[string]interface{}
+}
+
+func NewDatabaseCmd(item *BeefyItem, cmdType CmdType, instructions map[string]interface{}) DatabaseCmd {
+	return DatabaseCmd{
+		Item:         item,
+		Type:         cmdType,
+		Instructions: instructions,
+	}
+}
+
+type Database struct {
+	DB       *gorm.DB
+	messages <-chan DatabaseCmd
+	log      *logrus.Entry
+}
+
+func NewDatabase(db *gorm.DB, messages <-chan DatabaseCmd, log *logrus.Entry) *Database {
+	return &Database{
+		DB:       db,
+		messages: messages,
+		log:      log,
+	}
+}
+
+func PrepareDatabase(config *Config) (*gorm.DB, error) {
+	config.DBConfig.DBPath = "tmp.db"
+	tmpDBFile, err := ioutil.TempFile("", config.DBConfig.DBPath)
+	if err != nil {
+		return nil, err
+	}
+
+	db, err := gorm.Open(config.DBConfig.Dialect, tmpDBFile.Name())
+	if err != nil {
+		return nil, err
+	}
+
+	InitTables(db)
+
+	return db, nil
+}
+
+func InitTables(db *gorm.DB) {
+	var beefyItem BeefyItem
+	if !db.HasTable(&beefyItem) {
+		db.CreateTable(&beefyItem)
+		db.Model(&beefyItem)
+	}
+}
+
+func (d *Database) onDone(ctx context.Context) error {
+	d.log.Info("Shutting down database...")
+	// close(d.messages) TODO: properly close channels
```

Comment:
do in this pr plz

Reply:
Updated relaychain listener to shut down the beefy channel used by the database write loop in https://github.com/Snowfork/polkadot-ethereum/pull/314/commits/f03c815ec36c601637c07e7ca1695ce85ab6e85a.